File: server/internal/handler/call.go
package handler

import (
	"net/http"

	"github.com/gin-gonic/gin"
	"github.com/google/uuid"
	"github.com/yuanchat/server/internal/middleware"
	"github.com/yuanchat/server/internal/repository"
	"github.com/yuanchat/server/internal/service"
	"go.uber.org/zap"
)

// CallHandler 通话相关的 REST 端点。
//
// 通话本身全部走 WebSocket 信令；这里只有两个「拿配置」的读端点。
type CallHandler struct {
	calls  *service.CallService
	users  *repository.UserRepository
	logger *zap.Logger
}

// NewCallHandler 构造通话 handler。
func NewCallHandler(
	calls *service.CallService, users *repository.UserRepository, logger *zap.Logger,
) *CallHandler {
	return &CallHandler{calls: calls, users: users, logger: logger}
}

// callParticipantDTO 房间快照里的参与者（比 WS 帧多不了字段，但两处结构独立演进）。
type callParticipantDTO struct {
	UserID    uuid.UUID `json:"user_id"`
	ConnID    string    `json:"conn_id"`
	Nickname  string    `json:"nickname"`
	AvatarURL *string   `json:"avatar_url,omitempty"`
	State     string    `json:"state"`
}

// callRoomDTO 房间快照响应体。
type callRoomDTO struct {
	CallID         uuid.UUID            `json:"call_id"`
	ConversationID uuid.UUID            `json:"conversation_id"`
	Media          string               `json:"media"`
	State          string               `json:"state"`
	CallerID       uuid.UUID            `json:"caller_id"`
	Participants   []callParticipantDTO `json:"participants"`
}

// GetICEServers 下发 STUN/TURN 配置与临时凭据。
//
// @Summary 获取 ICE 服务器配置
// @Tags calls
// @Success 200 {object} Response
// @Router /calls/ice-servers [get]
func (h *CallHandler) GetICEServers(c *gin.Context) {
	userID, ok := middleware.GetUserID(c)
	if !ok {
		Unauthorized(c, "unauthorized")
		return
	}
	Success(c, h.calls.ICEServers(userID))
}

// GetCall 返回通话房间快照。
//
// 唯一调用方是桌面端的独立通话窗口：它是在 call.incoming 之后才被创建的，
// 自己的 WebSocket 连上时那一帧早已发完，没有这个端点就没有房间信息可渲染。
// 顺带让「通话窗口刷新 / 重开」天然可恢复。
//
// @Summary 获取通话房间快照
// @Tags calls
// @Param call_id path string true "通话 ID"
// @Success 200 {object} Response
// @Router /calls/{call_id} [get]
func (h *CallHandler) GetCall(c *gin.Context) {
	userID, ok := middleware.GetUserID(c)
	if !ok {
		Unauthorized(c, "unauthorized")
		return
	}
	callID, err := uuid.Parse(c.Param("call_id"))
	if err != nil {
		BadRequest(c, "invalid call id")
		return
	}

	room, err := h.calls.Get(c.Request.Context(), callID)
	if err != nil {
		NotFound(c, "call not found")
		return
	}

	// 只有房间参与者能读快照：里面含各方的 conn_id，那是信令的定址凭据，
	// 泄露给外人等于把「往任意参与者连接上塞帧」的门票发出去
	inRoom := false
	ids := make([]uuid.UUID, 0, len(room.Participants))
	for _, p := range room.Participants {
		ids = append(ids, p.UserID)
		if p.UserID == userID {
			inRoom = true
		}
	}
	if !inRoom {
		Error(c, http.StatusForbidden, http.StatusForbidden, "not a call participant")
		return
	}

	nick := map[uuid.UUID]string{}
	avatar := map[uuid.UUID]*string{}
	// 昵称查不到不影响通话本身：快照的关键字段是 conn_id 与 state，
	// 与「查询失败就留空」同一姿态，不因一次用户表查询挂掉整个房间快照
	if h.users != nil {
		if users, err := h.users.FindByIDs(c.Request.Context(), ids); err == nil {
			for i := range users {
				nick[users[i].ID] = users[i].Nickname
				avatar[users[i].ID] = users[i].AvatarURL
			}
		} else {
			h.logger.Warn("取通话参与者资料失败", zap.Error(err))
		}
	}

	out := callRoomDTO{
		CallID:         room.CallID,
		ConversationID: room.ConversationID,
		Media:          string(room.Media),
		State:          string(room.State),
		CallerID:       room.CallerID,
		Participants:   make([]callParticipantDTO, 0, len(room.Participants)),
	}
	for _, p := range room.Participants {
		out.Participants = append(out.Participants, callParticipantDTO{
			UserID:    p.UserID,
			ConnID:    p.ConnID,
			Nickname:  nick[p.UserID],
			AvatarURL: avatar[p.UserID],
			State:     string(p.State),
		})
	}
	Success(c, out)
}

File: packages/shared/src/api/call.ts
/**
 * 通话 REST 客户端 — ICE 凭据签发与房间快照
 *
 * @description
 * 两个端点各只有一个调用方：
 * - `GET /calls/ice-servers`：建 `RTCPeerConnection` 前取 STUN/TURN 凭据
 *   （TURN 凭据是 HMAC 时效凭据，不能写死在前端，故必须走接口）
 * - `GET /calls/:call_id`：桌面通话窗口启动时拉快照 —— 窗口是在 `call.incoming`
 *   之后才创建的，它自己的 WS 连上时那一帧早已发完，不拉快照就没有房间信息可渲染
 */
import { apiGet } from "./client";
import type { CallMedia, CallParticipant } from "../ws/chatSocket";

/** ICE 凭据响应（与 `server/internal/service/call_service.go` 的 `ICEConfig` 对齐）。 */
export interface CallIceConfigDTO {
  ice_servers: RTCIceServer[];
  /** 凭据有效期（秒），客户端据此缓存 */
  ttl: number;
}

/**
 * 房间快照响应。
 *
 * @remarks 字段逐字对齐 `server/internal/handler/call.go` 的 `callRoomDTO` ——
 *   注意它给的是 `caller_id` 而不是主叫的完整资料：主叫昵称/头像已经随
 *   `call.incoming` / `call.state` 的 `participants` 一起下发过，快照不重复带。
 */
export interface CallRoomDTO {
  call_id: string;
  conversation_id: string;
  media: CallMedia;
  state: "ringing" | "active";
  caller_id: string;
  participants: CallParticipant[];
}

/**
 * 拉取 ICE 服务器列表与凭据有效期。
 *
 * @remarks 带 TTL 缓存的封装在 `webrtc/iceServers.ts`，业务侧应调那一个 ——
 *   本函数每次都真的打接口。
 */
export function fetchIceConfig(): Promise<CallIceConfigDTO> {
  return apiGet<CallIceConfigDTO>("/api/v1/calls/ice-servers");
}

/**
 * 拉取通话房间当前快照。
 *
 * @param callId - 通话 id
 * @throws ApiError code=403 非房间参与者 / code=404 房间不存在（已终结）
 */
export function fetchCall(callId: string): Promise<CallRoomDTO> {
  return apiGet<CallRoomDTO>("/api/v1/calls/" + encodeURIComponent(callId));
}
